File: 2 - Regression/2b - Multiple Linear Regression/multiple-linear-regression.py
#	Multiple Linear Regression 

#Importing the libraries
import numpy as np 
import matplotlib.pyplot as plt
import pandas as pd 

#Import the dataset
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[: , :-1].values
y = dataset.iloc[: , -1].values


#Encoding Catagorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:,-1] = labelencoder_X.fit_transform(X[:,-1])
onehotencoder = OneHotEncoder(categorical_features=[-1])
X = onehotencoder.fit_transform(X).toarray()

# Avoiding the dummy variable trap
X = X[:, 1:]

#Splitting the dataset into training and test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size = 0.2, random_state = 0)

#Fitting Multiple Linear Regression to the Training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train,y_train)

#Predicting test set results
y_pred = regressor.predict(X_test)


#==================================================================#
#   Backward Elimination Method for Multiple Linear Regression     #
#==================================================================#

#   Building the optimal model using Backward Elimination manually
"""import statsmodels.formula.api as sm
X = np.append(arr = np.ones((50,1)).astype(int), values= X, axis = 1)
X_opt = X[:,[0,1,2,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
#Removing x2 having p value 0.990
X_opt = X[:,[0,1,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
#Removing x1 having p value 0.940
X_opt = X[:,[0,3,4,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
#Removing x2 having p value 0.602
X_opt = X[:,[0,3,5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()
#Removing x2 having p value 0.060
X_opt = X[:,[0,3]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
regressor_OLS.summary()"""


#   Building the optimal model using Backward Elimination automatically
from statsmodels.tools.tools import add_constant
X_opt = add_constant(X,prepend = True)
import statsmodels.formula.api as sm
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
initial_summary = regressor_OLS.summary()
pvalues = regressor_OLS.pvalues
rsquared_adj = regressor_OLS.rsquared_adj


while max(pvalues) > 0.05:
    max_index = np.argmax(pvalues)
    X_opt = np.delete(X_opt, max_index, axis=1)
    new_regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()
    
    if rsquared_adj > new_regressor_OLS.rsquared_adj:
        break
   
    regressor_OLS = new_regressor_OLS
    pvalues = regressor_OLS.pvalues
    rsquared_adj = regressor_OLS.rsquared_adj

final_summary = regressor_OLS.summary()

#Splitting the dataset into training and test set
X_opt_train, X_opt_test, y_opt_train, y_opt_test = train_test_split(X_opt,y, test_size = 0.2, random_state = 0)

#Fitting Multiple Linear Regression to the Training set
regressor_OLS = LinearRegression()
regressor_OLS.fit(X_opt_train,y_opt_train)

#Predicting test set results
y_opt_pred = regressor_OLS.predict(X_opt_test)

File: 2 - Regression/2b - Multiple Linear Regression/multiple-linear-regression.R
# Multiple Linear Regression

#Import the dataset
dataset = read.csv('50_Startups.csv')

#Encoding Catagorical data
dataset$State =  factor(dataset$State,
                        levels = c('New York','California','Florida'),
                        labels = c(1,2,3))

#Splitting the dataset into training and test set
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit,SplitRatio = 0.8)
training_set = subset(dataset,split == TRUE)
test_set = subset(dataset,split == FALSE)

# Fitting Multiple Linear Regression to the Training set
#regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State ,
#               data = training_set)
regressor = lm(formula = Profit ~ . ,
               data = training_set)
summary(regressor)

# Predicting Test Set results
y_pred = predict(regressor,newdata = test_set)


#===========================================================#
#   Building the optimal model using Backward Elimination   #
#===========================================================#
# Manually building optimum regressor
regressor_optimum = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State ,
              data = dataset)
summary(regressor_optimum)
# Removing State2 & State3
regressor_optimum = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,
               data = dataset)
summary(regressor_optimum)
# Removing Administration
regressor_optimum = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,
               data = dataset)
summary(regressor_optimum)
# Removing Marketing Spend
regressor_optimum = lm(formula = Profit ~ R.D.Spend ,
               data = dataset)
summary(regressor_optimum)

# Automatically building optimum regressor using AIC instead of p-value
    # regressor_optimum = lm(formula = Profit~.,
    #                        data = dataset)
    # initial_summary = summary(regressor_optimum)
    # regressor_optimum = step(regressor_optimum, direction = "backward")
    # final_summary = summary(regressor_optimum)


# Predicting Test Set results
y_opt_pred = predict(regressor_optimum,newdata = test_set)












